File: bht.py
import time
import sys
from utils import utils
from utils.utils import Contador
from flask import Flask, render_template


M = 10 # Número de bits do PC usados para indexar a BHT
N = 4  # Número de bits em cada registrador da BHT


# Lendo o arquivo de trace e obtendo uma lista de tuplas na forma (endereço do desvio, direcao realizada)
desvios = utils.ler_arquivo_trace('traces/trace1.txt', M)


# Cria e inicializa a BHT
BHT = {}
for endereco, direcao in desvios:
    if endereco not in BHT:
        BHT[endereco] = [0]*N # iniciando o registrador indexado por "endereço" com 0
        #print(endereco, BHT[endereco])


# Cria e inicializa os contadores
num_desvios = len(BHT)
contadores = {}
for endereco in BHT.keys():
    list_contadores = []
    for i in range(2**N):
        list_contadores.append(Contador())
    contadores[endereco] = list_contadores # Cada desvio terá 2^N contadores


def list_hist_desvios_to_int(list_hist_desvios):
    str_hist_desvios = "".join(map(str, list_hist_desvios))
    valor_int = int(str_hist_desvios, 2)
    return valor_int


def prever_direcao(endereco_desvio, BHT, contadores):
    list_hist_desvios = BHT[endereco_desvio]
    contador = contadores[endereco_desvio][list_hist_desvios_to_int(list_hist_desvios)]
    if contador.valor >= 2:
        return 't'
    return 'n'


def atualizar_tabelas(endereco_desvio, BHT, contadores, dir_realizada):
    BHT[endereco_desvio].insert(0, 1 if dir_realizada == 't' else 0)
    BHT[endereco_desvio].pop(-1)
    contador = contadores[endereco_desvio][list_hist_desvios_to_int(BHT[endereco_desvio])]
    if dir_realizada == 't':
        contador.incrementar()
    else:
        contador.decrementar()

app = Flask(__name__)


# Executa a simulação
for endereco, direcao in desvios:
    #time.sleep(.7)
    #print("Desvio:", endereco)
    #print("Histórico:", BHT[endereco])
    #print("Previsão:", prever_direcao(endereco, BHT, contadores))
    #time.sleep(.5)
    #print("Direção realizada:", direcao, "\n")
    atualizar_tabelas(endereco, BHT, contadores, direcao)


@app.route('/')
def index():
    contadores2 = {}
    for end, list_conts in contadores.items():
        list_valores = [cont.valor for cont in list_conts]
        contadores2[end] = list_valores
    return render_template('index.html', desvios=desvios, BHT=BHT, contadores2=contadores2)


if __name__ == "__main__":
    app.run(debug=True, port=5001)

File: src/js/bht.js
$(document).ready(function () {
    /* "desvios" armazenará os dados do arquivo de trace em um array de arrays onde cada 
    array elemento é da forma [endereço do desvio, direção do desvio] */
    var desvios = []

    // Lê o arquivo trace selecionado pelo usuário e preenche o Array "desvios"
    $('#arqTrace').change(function(e){
            arquivo = e.target.files[0]
            fr = new FileReader()
            fr.onload = function(){
                linhas = fr.result.split('\n')

                // Remove a última linha, se ela for vazia
                if(linhas[linhas.length - 1] == '')
                    linhas.pop()

                for(i = 0; i < linhas.length; ++i){
                    var [enderDesvio, direcao] = linhas[i].split(' ')
                    desvios.push([enderDesvio, direcao])
                }
            }
            fr.readAsText(arquivo)
        })
    
    // Seta o valor máximo de n para o valor de m, quando m for setado
    $('#m').change(function(){
        $('#n').attr('max', $('#m').val())
    })

    var m, n, BHT, PHT
    // Inicia a simulação quando o botão "Iniciar simulação" for pressionado
    $('#btn-iniSim').click(function(){
        // Desativa alterações em m e n e a entrada de arquivo
        desativarEntradas()
        
        // inicializa as variáveis m e n com os valores selecionados pelo usuário
        m = $('#m').val()
        n = $('#n').val()
        
        carregarTabelaDesvios(desvios) // Exibe os endereços na tabela de desvios 
        BHT = carregarBHT(desvios, m, n) // Exibe o estado inicial da BHT
        PHT = carregarPHT(BHT, n) // Exibe o estado inicial da PHT
        
        // executa a simulação passo a passo
        simular(desvios, BHT, PHT, m, n)

        $(this).prop('disabled', true) // Desativa o botão "Iniciar simulação"
    })

    function simular(desvios, BHT, PHT, m, n){
        var i = 0, linhaBHTanterior, linhaPHTanterior, tdContadAnterior
        function loop(){
            setTimeout(function(){
                // Executa a previsão do i-ésimo desvio
                enderecoDesv = desvios[i][0]
                previsao = preverDesvio(enderecoDesv, BHT, PHT, m, i)
                
                // Atualiza a direção realizada
                dirRealizada = desvios[i][1]

                // Atuliza o painel de resultados
                atualizarPainelResults(previsao, dirRealizada, i)
                
                // Colore a tabela de desvios de acordo com a corretude da previsão
                atualizarTabDesvios(previsao, dirRealizada, i)

                // Remove a coloração da linha anterior da da BHT  
                linhaBHTatual = $('#BHT-'+enderecoHexToMFormat(enderecoDesv, m))
                if(i > 0 && linhaBHTatual != linhaBHTanterior){
                    linhaBHTanterior.css({'background-color': 'transparent'})
                }
                // Atualiza o histórico de desvios indexado por "enderDesvio" na "BHT"
                // de acordo com "dirRealizada"
                linhaBHTatual.css({'background-color': 'red'})
                atualizarBHT(BHT, enderecoDesv, dirRealizada, m)
                linhaBHTanterior = linhaBHTatual

                // Atuliza o contador indexado pelo histórico do desvio atual de acordo com "dirRealizada"
                tdContadAtual = atualizarPHT(PHT, enderecoDesv, dirRealizada, m)
                if(tdContadAnterior && tdContadAnterior != tdContadAtual)
                    tdContadAnterior.css({'background-color': 'transparent'})
                tdContadAnterior = tdContadAtual
                linhaPHTatual = $('#PHT-'+enderecoHexToMFormat(enderecoDesv, m))
                linhaPHTatual.css({'background-color': 'gray'})
                if(linhaPHTanterior && linhaPHTatual != linhaPHTanterior)
                    linhaPHTanterior.css({'background-color': 'transparent'})
                linhaPHTanterior = linhaPHTatual

                ++i
                if(i < desvios.length){
                    loop()
                }
            }, 500)
        }
        loop()
    }

    function atualizarBHT(BHT, enderDesvio, dirRealizada, m){
        // Atualiza o histórico de desvios indexado por "enderDesvio" na "BHT"
        // de acordo com "dirRealizada"
        indiceBHT = enderecoHexToMFormat(enderecoDesv, m)
        if(dirRealizada.toLowerCase() == 't')
            BHT[indiceBHT].unshift(1)
        else
            BHT[indiceBHT].unshift(0)
        BHT[indiceBHT].pop()
        $('#hist-'+indiceBHT).text(BHT[indiceBHT])
    }

    function atualizarPHT(PHT, enderDesvio, dirRealizada, m){
        enderDesvMFormat = enderecoHexToMFormat(enderDesvio, m)
        contadores = PHT[enderDesvMFormat]
        strHistDesv = BHT[enderDesvMFormat].join('')
        indiceContador = historicoToIndice(strHistDesv)
        contador = contadores[indiceContador]
        tdContad = $('#conts-'+enderDesvMFormat+'-'+indiceContador)
        if(dirRealizada.toLowerCase() == 't')
            contador.incrementar()
        else
            contador.decrementar()
        tdContad.css({'background-color': 'red'})
        tdContad.text(contador.valor)
        return tdContad
    }
    
    function enderecoHexToMFormat(strEnderDesv, m){
        // Converte um endereço de desvio em hexadecimal para os m bits LSB 
        // excluindo-se os dois bits menos significativos
        enderBin32B = hexToBin32(strEnderDesv)
        return enderBin32B.substring(32 - m - 2, 32 - 2)
    }

    function preverDesvio(enderecoDesvio, BHT, PHT, m, i){
        enderInMFormat = enderecoHexToMFormat(enderecoDesvio, m)
        histDoDesvio = BHT[enderInMFormat]
        histDesvStr = histDoDesvio.join('')
        indiceContad = historicoToIndice(histDesvStr)
        contadores = PHT[enderInMFormat]
        contador = contadores[indiceContad]
        if(contador.valor >= 2)
            return 'T'
        
        else
            return 'N'
        
    }

    function carregarBHT(desvios, m, n){
        
        // Cria e inicializa um objeto BHT no formato {endereço desvio: histtórico}
        BHT = {}
        tbodyTabBHT = $('#tabela-bht').children('tbody')
        for(i = 0; i < desvios.length; ++i){
            enderDesvioHex = desvios[i][0]
            // converte para binário e mantém apenas o m LSB sem considerar os 2 bits menos significativos
            enderLinhaBHT = hexToBin32(enderDesvioHex).substring(32 - m - 2, 32 - 2)
            if(!(enderLinhaBHT in BHT)){
                historico = new Array(parseInt(n)).fill(0) // cria um objeto Array contendo n zeros
                BHT[enderLinhaBHT] = historico
                tbodyTabBHT.append('<tr id="BHT-'+enderLinhaBHT+'" name="linha-BHT">\
                    <td>'+enderLinhaBHT+'</td><td id="hist-'+enderLinhaBHT+'">'+historico+'</td></tr>')
            }
        }
        return BHT
    }

    function carregarPHT(BHT, n){
        enderecosBHT = Object.keys(BHT)
        PHT = {}
        for(i = 0; i < enderecosBHT.length; ++i){
            PHT[enderecosBHT[i]] = []
            for(j = 0; j < Math.pow(2, n); ++j){
                PHT[enderecosBHT[i]].push(new Contador())
            }
        }
        tbodyTabPHT = $("#tabela-pht").children('tbody')
        for(i = 0; i < enderecosBHT.length; ++i){
            tbodyTabPHT.append('<tr id="PHT-'+enderecosBHT[i]+'" name="linha-PHT">\
                <td>'+enderecosBHT[i]+'</td></tr>')
            linhaPHT = $('#PHT-'+enderecosBHT[i])
            contadores = PHT[enderecosBHT[i]]
            for(j = 0; j < contadores.length; ++j){
                linhaPHT.append('<td id="conts-'+enderecosBHT[i]+'-'+j.toString()+'"\
                    class="contadores">'+contadores[j].valor+'</td>')
            }
        }
        /* Seta o colspan do título "contadores"*/
        $("#th-contadores").attr('colspan', Math.pow(2, n).toString())
        return PHT
    }
 
})